python/src/interpreter.h

```diff
@@ -0,0 +1,307 @@
+#include <iostream>
+#include <map>
+#include <memory>
+#include <pybind11/numpy.h>
+#include <pybind11/pybind11.h>
+#include <type_traits>
+
+namespace triton {
+
+namespace interpreter {
+
+enum class MemSemantic { ACQUIRE_RELEASE, ACQUIRE, RELEASE, RELAXED };
+
+enum class RMWOp { ADD, FADD, AND, OR, XOR, XCHG, MAX, MIN, UMIN, UMAX };
+
+std::map<MemSemantic, int> mem_semantic_map = {
+    {MemSemantic::ACQUIRE_RELEASE, __ATOMIC_ACQ_REL},
+    {MemSemantic::ACQUIRE, __ATOMIC_ACQUIRE},
+    {MemSemantic::RELEASE, __ATOMIC_RELEASE},
+    {MemSemantic::RELAXED, __ATOMIC_RELAXED},
+};
+
+// Use compiler builtin atomics instead of std::atomic which requires
+// each variable to be declared as atomic.
+// Currently work for clang and gcc.
+template <bool is_min, typename T> T atomic_cmp(T *ptr, T val, int order) {
+  auto cmp = [](T old, T val) {
+    if constexpr (is_min) {
+      return old > val;
+    } else {
+      return old < val;
+    }
+  };
+  // First load
+  T old_val = __atomic_load_n(ptr, order);
+  while (cmp(old_val, val)) {
+    if (__atomic_compare_exchange(ptr, &old_val, &val, false, order, order)) {
+      break;
+    }
+  }
+  return old_val;
+}
```

Comment:
any reason all those templates need to be in the header? Looks like they are only used in the cpp file so they could be moved there?

Reply:
Sure, I'm just afraid of the cpp file getting too long. I'll move them back to the cpp file

Reply:
Fair point, maybe we can split it up into 2 cpp but templates in header files are always scary to me ;)